pkg/webhooks/pod_webhook.go

```diff
@@ -113,30 +120,65 @@ func (p *PodWebhook) Default(ctx context.Context, obj runtime.Object) error {
+		if (foundEpKey) && !exclusiveAffinityApplied(*pod, epKey) {
+			SetExclusiveAffinities(pod, groupUniqueKey, epKey, leaderworkerset.GroupUniqueHashLabelKey)
+		}
+		_, foundSubGroupSize := pod.Annotations[leaderworkerset.SubGroupSizeAnnotationKey]
+		if foundSubGroupSize && pod.Labels[leaderworkerset.SubGroupIndexLabelKey] == "" {
+			// The leader pod will always lands on SubGroup 0 in the same
+			// topology regardless if it requests any accelerator resources
+			pod.Labels[leaderworkerset.SubGroupIndexLabelKey] = "0"
+			subGroupUniqueKey := genGroupUniqueKey(pod.Name, "0")
+			pod.Labels[leaderworkerset.SubGroupUniqueHashLabelKey] = subGroupUniqueKey
+			subEpKey, foundSubEpKey := pod.Annotations[leaderworkerset.SubGroupExclusiveKeyAnnotationKey]
+			if foundSubEpKey && !exclusiveAffinityApplied(*pod, subEpKey) {
+				SetExclusiveAffinities(pod, subGroupUniqueKey, subEpKey, leaderworkerset.SubGroupUniqueHashLabelKey)
+			}
 		}
 	} else {
 		_, workerIndex := statefulsetutils.GetParentNameAndOrdinal(pod.Name)
 		if workerIndex == -1 {
 			return fmt.Errorf("parsing pod ordinal for pod %s", pod.Name)
 		}
 		pod.Labels[leaderworkerset.WorkerIndexLabelKey] = fmt.Sprint(workerIndex)
+		subGroupSize, foundSubGroupSize := pod.Annotations[leaderworkerset.SubGroupSizeAnnotationKey]
+		if foundSubGroupSize && pod.Labels[leaderworkerset.SubGroupIndexLabelKey] == "" {
+			subGroupSizeInt, err := strconv.Atoi(subGroupSize)
+			if err != nil {
+				return err
+			}
+			var subGroupIndexKey string
+			if (podCount-1)%subGroupSizeInt == 0 {
+				// Leader is considered as extra pod, it is part of the first group
+				subGroupIndexKey = fmt.Sprint((workerIndex - 1) / subGroupSizeInt)
+				pod.Labels[leaderworkerset.SubGroupIndexLabelKey] = subGroupIndexKey
+			} else {
+				subGroupIndexKey = fmt.Sprint(workerIndex / subGroupSizeInt)
+				pod.Labels[leaderworkerset.SubGroupIndexLabelKey] = subGroupIndexKey
+			}
+			leaderName := pod.Annotations[leaderworkerset.LeaderPodNameAnnotationKey]
+			subGroupUniqueKey := genGroupUniqueKey(leaderName, subGroupIndexKey)
+			pod.Labels[leaderworkerset.SubGroupUniqueHashLabelKey] = subGroupUniqueKey
+			subEpKey, foundSubEpKey := pod.Annotations[leaderworkerset.SubGroupExclusiveKeyAnnotationKey]
+			if foundSubEpKey && !exclusiveAffinityApplied(*pod, subEpKey) {
+				SetExclusiveAffinities(pod, subGroupUniqueKey, subEpKey, leaderworkerset.SubGroupUniqueHashLabelKey)
+			}
+		}
 	}
 
 	// injecting env vars if needed
 	if acceleratorutils.PodRequestsTPUs(pod.Spec) {
-		size, exist := pod.Annotations[leaderworkerset.SizeAnnotationKey]
-		if !exist {
-			return fmt.Errorf("size annotation is unexpectedly missing for pod %s", pod.Name)
-		}
-		podCount, err := strconv.Atoi(size)
-		if err != nil {
-			return err
-		}
-		if err := acceleratorutils.AddTPUVariables(pod, podCount); err != nil {
-			return err
+
+		_, foundSubGroupSize := pod.Annotations[leaderworkerset.SubGroupSizeAnnotationKey]
+		if foundSubGroupSize {
```

Comment:
Move this if/else statement to inside `AddTPUVariables`; this is a detail related to TPUs.

Reply:
done